### PR TITLE
Adding TTD support for generators

### PR DESCRIPTION
**Pull request by mike-kaufman**
Generator functions weren't supported in TTD.  This PR adds support and a few tests.

### CODE REVIEW

File: lib/Runtime/Debug/TTSnapObjects.cpp [hunk 1]
```diff
@@ -2117,15 +2128,277 @@ namespace TTD
...
+            Js::Var* argVals = nullptr;
+            if (generatorInfo->arguments_count > 0) {
+                argVals = RecyclerNewArray(ctx->GetRecycler(), Field(Js::Var), generatorInfo->arguments_count);
+
+                for (Js::RegSlot i = 0; i < generatorInfo->arguments_count; i++) {
+                    argVals[i] = inflator->InflateTTDVar(generatorInfo->arguments_values[i]);
+                }
+            }
+
+            Js::ScriptFunction*  scriptFunction = nullptr;// static_cast<Js::ScriptFunction*>((generatorInfo->scriptFunction != TTD_INVALID_PTR_ID) ? inflator->LookupObject(generatorInfo->scriptFunction) : nullptr);
```
Comment (MSLaguana) [inline lib/Runtime/Debug/TTSnapObjects.cpp]: Is this commented out code related to the "todo" further down?

File: lib/Runtime/Debug/TTSnapObjects.cpp [hunk 2]
```diff
@@ -2117,15 +2128,277 @@ namespace TTD
...
+
+        void DoAddtlValueInstantiation_SnapGeneratorInfo(const SnapObject* snpObject, Js::RecyclableObject* obj, InflateMap* inflator) {
+            Js::ScriptContext* ctx = inflator->LookupScriptContext(snpObject->SnapType->ScriptContextLogId);
+            Js::JavascriptGenerator* generator = reinterpret_cast<Js::JavascriptGenerator*>(obj);
+
+            SnapGeneratorInfo* generatorInfo = SnapObjectGetAddtlInfoAs<SnapGeneratorInfo*, SnapObjectType::SnapGenerator>(snpObject);
+            Js::ScriptFunction* scriptFunction = reinterpret_cast<Js::ScriptFunction*>((generatorInfo->scriptFunction != TTD_INVALID_PTR_ID) ? inflator->LookupObject(generatorInfo->scriptFunction) : nullptr);
+            generator->SetScriptFunction(scriptFunction);
+
+            Js::FunctionBody* executeFunction = Js::JavascriptFunction::UnsafeFromVar(scriptFunction)->GetFunctionBody();
```
Comment (MSLaguana) [inline lib/Runtime/Debug/TTSnapObjects.cpp]: Why is the `UnsafeFromVar` required here? Doesn't `Js::ScriptFunction` inherit from `Js::JavascriptFunction` and so `GetFunctionBody` should already be a member?

Comment (MSLaguana) [reply lib/Runtime/Debug/TTSnapObjects.cpp]: And I just saw that `scriptFunction` can be `nullptr` if `generatorInfo->scriptFunction` is `TTD_INVALID`; is that a recoverable state, or should we failfast?

File: lib/Runtime/Debug/TTSnapObjects.cpp [hunk 3]
```diff
@@ -2117,15 +2128,277 @@ namespace TTD
...
+
+            Js::Var* frameSlotArray = nullptr;
+            if (generatorInfo->frame_slotCount != 0) {
+                frameSlotArray = RecyclerNewArray(ctx->GetRecycler(), Field(Js::Var), generatorInfo->frame_slotCount);
+
+                for (Js::RegSlot i = 0; i < generatorInfo->frame_slotCount; i++) {
+                    if (generatorInfo->frame_slotArray[i] != nullptr) {
+                        frameSlotArray[i] = inflator->InflateTTDVar(generatorInfo->frame_slotArray[i]);
+                    }
+                    else {
```
Comment (MSLaguana) [inline lib/Runtime/Debug/TTSnapObjects.cpp]: nit: braces should be on the following line (this applies to all the braces you've added/changed)

File: lib/Runtime/Debug/TTSnapObjects.cpp [hunk 4]
```diff
@@ -2117,15 +2128,277 @@ namespace TTD
...
+            }
+            reader->ReadSequenceEnd();
+
+            sgi->byteCodeReader_offset = reader->ReadUInt32(NSTokens::Key::u32Val, true);
+
+            SnapObjectSetAddtlInfoAs<SnapGeneratorInfo*, SnapObjectType::SnapGenerator>(snpObject, sgi);
+        }
+
+#if ENABLE_SNAPSHOT_COMPARE
+        void AssertSnapEquiv_SnapGeneratorInfo(const SnapObject *sobj1, const SnapObject *sobj2, TTDCompareMap &compareMap) {
```
Comment (MSLaguana) [inline lib/Runtime/Debug/TTSnapObjects.cpp]: Looks like this needs a "todo" to be implemented, just making sure that mappings between the objects are consistent?

File: lib/Runtime/Debug/TTSnapObjects.h [hunk 5]
```diff
@@ -1075,6 +1080,70 @@ namespace TTD
...
+
+        };
+       
+        struct SnapGeneratorInfo {
+            int generatorPrototype; // 0 == unknown, 1 == nullType, 2 == generatorPrototype
+            Js::RegSlot frame_slotCount;
+            TTDVar* frame_slotArray;
+            uint32 state; // enum value of JavascriptGenerator.GeneratorState
+            TTD_PTR_ID scriptFunction;
+            byte arguments_callInfo[sizeof(Js::CallInfo)];
```
Comment (MSLaguana) [inline lib/Runtime/Debug/TTSnapObjects.h]: Just checking: while we do have a goal of windows/xplat trace compatibility, we don't have a goal of cross-architecture compatibility, right? So potential endianness issues aren't important?

File: lib/Runtime/Language/InterpreterStackFrame.cpp [hunk 6]
```diff
@@ -1746,6 +1746,46 @@ namespace Js
...
+        //
+        // Allocate a new InterpreterStackFrame instance on the recycler heap.
+        // It will live with the JavascriptGenerator object.
+        //
+        ScriptContext* functionScriptContext = function->GetScriptContext();
+        Arguments generatorArgs = generator->GetArguments();
+        InterpreterStackFrame::Setup setup(function, generatorArgs);
+        size_t varAllocCount = setup.GetAllocationVarCount();
+        size_t varSizeInBytes = varAllocCount * sizeof(Var);
+        DWORD_PTR stackAddr = reinterpret_cast<DWORD_PTR>(&generator); // as mentioned above, use any stack address from this frame to ensure correct debugging functionality
```
Comment (MSLaguana) [inline lib/Runtime/Language/InterpreterStackFrame.cpp]: I see that this code was outlined from further down, but the comment on this line seem to have lost context

File: lib/Runtime/Library/JavascriptGenerator.cpp [hunk 7]
```diff
@@ -232,14 +249,124 @@ namespace Js
...
+
+        // grab arguments
+        TTD_PTR_ID* depArray = nullptr;
+        uint32 depCount = 0;
+        gi->arguments_count = this->args.GetArgCountWithExtraArgs();
+        gi->arguments_values = nullptr;
+        if (gi->arguments_count > 0)
+        { 
+            gi->arguments_values = alloc.SlabAllocateArray<TTD::TTDVar>(gi->arguments_count);
+            depArray = alloc.SlabReserveArraySpace<TTD_PTR_ID>(gi->frame_slotCount + gi->arguments_count);
```
Comment (MSLaguana) [inline lib/Runtime/Library/JavascriptGenerator.cpp]: I'm a bit confused here; why do we "reserve but not commit" space for `depArray`, then write to it in the next for loop (why is that permitted if we have not committed the allocation?) before committing the allocation? Or is this reserve/commit notion different to the standard memory reserve/commit notion? E.g. reserving a memory allocation would typically block out a section of virtual memory without commiting physical memory for it, but you can't touch the memory until you commit it, right?

File: test/TTBasic/generatorBasic.js [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+    yield i++;
+    yield i++;
+    yield i++;
+    yield i++;
+}
+
+var gen = testGenerator();
+
+function yieldOne() {
+    var v1 = gen.next();
```
Comment (MSLaguana) [inline test/TTBasic/generatorBasic.js]: Could probably do with some tests that make use of the fact that the `yield` expression in a generator can be given a value by `gen.next(val)`, and make sure that is captured correctly as well. 

can/should we also force some TTD snapshots mid-generator execution?

File: lib/Runtime/Debug/TTSnapObjects.cpp [hunk 9]
```diff
@@ -779,12 +789,15 @@ namespace TTD
             writer->WriteBool(NSTokens::Key::boolVal, snapFuncInfo->HasSuperReference, NSTokens::Separator::CommaSeparator);
         }
 
-        void ParseAddtlInfo_SnapScriptFunctionInfo(SnapObject* snpObject, FileReader* reader, SlabAllocator& alloc)
-        {
```
Comment (mrkmarron) [inline lib/Runtime/Debug/TTSnapObjects.cpp]: It looks like some braces/whitespace was changed here (and elsewhere). We need to get it back into the correct ChakraCore style.

File: lib/Runtime/Debug/TTSnapObjects.h (hunk 5 above)
Comment (mrkmarron) [reply lib/Runtime/Debug/TTSnapObjects.h]: That is correct, cross-architecture is explicitly not supported.

File: lib/Runtime/Library/JavascriptGenerator.cpp (hunk 7 above)
Comment (mrkmarron) [reply lib/Runtime/Library/JavascriptGenerator.cpp]: We use a simple bump pointer slab allocator for the snapshots. The SlabReserve(k) call ensures there are at least k contiguous elements available to write into while SlabCommit(k) actually does the bump pointer operation to commit them. This way we don't need to do a bump pointer bounds check every iteration.

Comment (mrkmarron) [inline lib/Runtime/Library/JavascriptGenerator.cpp]: "gi->frame_slotCount + gi->arguments_count" is too large since you are not writing any localSlots into the depArray. I think this should just be "gi->arguments_count"

File: test/TTBasic/generatorBasicReplay.baseline [hunk 10]
```diff
@@ -0,0 +1,3 @@
+gen.next() = {value: 2, done: false}
+
```
Comment (mrkmarron) [inline test/TTBasic/generatorBasicReplay.baseline]: Shouldn't this also print "gen.next() = {value: 3, done: false}"?

Comment (mike-kaufman) [reply test/TTBasic/generatorBasicReplay.baseline]: yes.  I don't exactly understand why - was thinking it was because the state of that last `WScript.SetTimeout(...)` call was lost, but need to confirm

File: lib/Runtime/Debug/TTSnapObjects.cpp (hunk 3 above)
Comment (mike-kaufman) [reply lib/Runtime/Debug/TTSnapObjects.cpp]: should be fixed now.

File: lib/Runtime/Debug/TTSnapObjects.cpp (hunk 9 above)
Comment (mike-kaufman) [reply lib/Runtime/Debug/TTSnapObjects.cpp]: should be fixed now.

File: lib/Runtime/Debug/TTSnapObjects.cpp (hunk 1 above)
Comment (mike-kaufman) [reply lib/Runtime/Debug/TTSnapObjects.cpp]: this was dead code.  I cleaned it up since we set the `scriptFunction` after object instantiation.

File: lib/Runtime/Debug/TTSnapObjects.cpp (hunk 4 above)
Comment (mike-kaufman) [reply lib/Runtime/Debug/TTSnapObjects.cpp]: fixed

File: lib/Runtime/Debug/TTSnapObjects.cpp (hunk 2 above)
Comment (mike-kaufman) [reply lib/Runtime/Debug/TTSnapObjects.cpp]: fixed - removed `UnsafeFromVar` call & added an `AssertOrFailFast` call if scriptFunction is null.

File: lib/Runtime/Language/InterpreterStackFrame.cpp (hunk 6 above)
Comment (mike-kaufman) [reply lib/Runtime/Language/InterpreterStackFrame.cpp]: fixed up comment.

File: lib/Runtime/Library/JavascriptGenerator.cpp (hunk 7 above)
Comment (mike-kaufman) [reply lib/Runtime/Library/JavascriptGenerator.cpp]: yes, fixed.

File: test/TTBasic/generatorBasic.js (hunk 8 above)
Comment (mike-kaufman) [reply test/TTBasic/generatorBasic.js]: ok, added additional test cases that verify the following:
  - snapshot of a completed generator
  - restore of a completed generator
  - calling writeTTDLog during generator execution
  - restoring at different log points and esnuring expected behavior based on that log point.

**Comment by mike-kaufman** [comment]
OK, I think I addressed all of the initial feedback.  Plz have another look & lmk if I missed anything.

File: test/TTBasic/generatorBasic.js (hunk 8 above)
Comment (MSLaguana) [reply test/TTBasic/generatorBasic.js]: Just took a look and I don't see any tests for something like

```js
function* gen() {
  var x = yield 1;
  return x;
}

var it = gen();
console.log(it.next());
console.log(it.next(2));
```

which should return values 1 and 2 respectively.

File: lib/Runtime/Debug/TTSnapObjects.cpp [hunk 11]
```diff
@@ -729,42 +731,51 @@ namespace TTD
                 func = ifunc;
             }
 
-            func->SetHasSuperReference(snapFuncInfo->HasSuperReference);
```
Comment (digitalinfinity) [inline lib/Runtime/Debug/TTSnapObjects.cpp]: Why is this removed?

File: lib/Runtime/Debug/TTSnapObjects.cpp [hunk 12]
```diff
@@ -2126,6 +2139,308 @@ namespace TTD
...
+            SnapGeneratorInfo* generatorInfo = SnapObjectGetAddtlInfoAs<SnapGeneratorInfo*, SnapObjectType::SnapGenerator>(snpObject);
+            Js::ScriptFunction* scriptFunction = reinterpret_cast<Js::ScriptFunction*>((generatorInfo->scriptFunction != TTD_INVALID_PTR_ID) ? inflator->LookupObject(generatorInfo->scriptFunction) : nullptr);
+            generator->SetScriptFunction(scriptFunction);
+
+            AssertOrFailFastMsg(scriptFunction != nullptr, "Nullptr for scriptFunction during DoAddtlValueInstantiation_SnapGeneratorInfo");
+
+            Js::FunctionBody* executeFunction = scriptFunction->GetFunctionBody();
+
+#if ENABLE_PROFILE_INFO
+            // call EnsureDynamicProfileInfo or we get asserts in some cases when createing the InterpreterStackFrame
```
Comment (digitalinfinity) [inline lib/Runtime/Debug/TTSnapObjects.cpp]: Nit: s/createing/creating

File: lib/Runtime/Debug/TTSnapObjects.cpp (hunk 11 above)
Comment (mike-kaufman) [reply lib/Runtime/Debug/TTSnapObjects.cpp]: good catch, this was a mistake.

File: test/TTBasic/generatorBasic.js (hunk 8 above)
Comment (mike-kaufman) [reply test/TTBasic/generatorBasic.js]: OK, added tests for this.  The tests will replay from all generated snap points, and on this branch, the last snap point fails.  This failure is fixed in the branch that contains async-await support, so leaving it commented out in here.

File: test/TTBasic/generatorBasicReplay.baseline (hunk 10 above)
Comment (mike-kaufman) [reply test/TTBasic/generatorBasicReplay.baseline]: this is fixed now.  THe issue was that we were calling `writeTTDLog` after last snapshot, and  when that happens, replay aborts.

File: lib/Runtime/Debug/TTSnapObjects.cpp (hunk 12 above)
Comment (mike-kaufman) [reply lib/Runtime/Debug/TTSnapObjects.cpp]: fixed

File: lib/Runtime/Debug/TTSnapObjects.cpp (hunk 11 above)
Comment (mike-kaufman) [reply lib/Runtime/Debug/TTSnapObjects.cpp]: fixed
